### DISCUSSION TITLE
Photos won't run without Google Play Serviecs

### DISCUSSION THREAD

**Discussion by sharshad1527** [ReVanced/revanced-discussions#1068] [Question]
Revanced Photos App, In Search Tab, In Your Map Option it's like a error **Photos won't run without Google Play Serviecs, which are not supported by your device** 
When i Click on it app crashes

my question is
is this a bug in revanced photos app?
is this a bug in GMS Core Service app?

**Comment by oSumAtrIX** [comment]
It is a bug by the GmsCore support patch.

**Reply by ak800i** [reply]
Not true, see my comment below. Please consider reopening the issue :)

**Reply by oSumAtrIX** [reply]
Not sure where its not true <img width="386" height="592" alt="image" src="https://github.com/user-attachments/assets/f1a268f8-dd4a-42f5-89b1-690169cd6a06" />

**Comment by ak800i** [comment]
## Root cause

Google Photos' bundled **Maps SDK** gates map initialization on a **signature check** of the Play Services provider:

- `bmbu.e()` (obfuscated `GooglePlayServicesUtilLight`) reads `app.revanced.android.gms`'s signing certificate (`getPackageInfo(..., GET_SIGNING_CERTIFICATES)`).
- It passes it to `bmbv.d()` (obfuscated `GoogleSignatureVerifier`), which compares against **Google's official cert** — SHA-1 `38918A453D07199354F8B19AF05EC6562CED5788` (found hardcoded in the app).
- GmsCore is **not** Google-signed (it's signed by ReVanced/your key), so the check fails → `a = false`.
- Final line: `return a || !"user".equals(Build.TYPE)`. Your device is a retail **`user`** build → returns **false** → the Maps SDK declares Play Services unavailable.
- Result: `MapsInitializer` **never loads the renderer**, so `BitmapDescriptorFactory` stays uninitialized, and `MapExplore.onStart()` throws **`IBitmapDescriptorFactory is not initialized`**.

**It is not fixable in GmsCore** (it can't be Google-signed). The fix belongs in **`ReVanced/revanced-patches`**.

## The fix (a ReVanced patch)

Extend the GmsCore-support patch to neutralize the Maps SDK signature check in Photos — i.e., force `bmbu.e()` (or `bmbv.d(PackageInfo, true)`) to accept GmsCore's signature. This is exactly analogous to the signature-spoofing the patch already does for the general GMS check; it just doesn't currently cover this Maps path.

**Reply by ak800i** [reply]
**Update — root cause confirmed and fixed on-device (correction to my earlier post).**

My earlier guess (`bmbu.e`/`bmbv.d`) was the right idea but the wrong method. Verified by decompiling + patching Photos and testing on a real device:

The actual gate is `GooglePlayServicesUtilLight.isGooglePlayServicesAvailable(Context, int)` (obfuscated `bmbu.b(Context,int)I`, called on the maps-init path (`MapsInitializer` → maps module loader). It requires **both** GMS version ≥ 13400000 and a Google-signature match on `app.revanced.android.gms`. GmsCore is not Google-signed, so it returns non-zero → the maps loader throws → `BitmapDescriptorFactory` is never initialized → `MapExplore.onStart()` crashes with `IBitmapDescriptorFactory is not initialized`.

Bumping the version alone does not fix it (the signature check still fails), and patching the sibling `isGooglePlayServicesAvailable(Context)` boolean / `GoogleSignatureVerifier` does not fix it (the init path uses the `(Context,int)` overload).

Fix (proven): neutralize `isGooglePlayServicesAvailable(Context,int)` to return `0` (SUCCESS) in Photos. The MapExplore feature then opens and shows geo-located photos. This belongs in the GmsCore-support patch. (Map tiles are a separate matter.)

**Reply by oSumAtrIX** [reply]
> Fix (proven): neutralize isGooglePlayServicesAvailable(Context,int) to return 0 (SUCCESS) in Photos. The MapExplore feature then opens and shows geo-located photos. This belongs in the GmsCore-support patch. (Map tiles are a separate matter.)

This should be doable fairly easy. Can you PR to https://gitlab.com/ReVanced/revanced-patches/-/blob/main/extensions/shared/library/src/main/java/app/revanced/extension/shared/GmsCoreSupport.java

**Reply by oSumAtrIX** [reply]
Also have you checked if maps starts working after that or if it just prevents the error from appearing

**Reply by ak800i** [reply]
Done opened 2 PRs:

- **Crash fix (this discussion):** revanced-patches MR → https://gitlab.com/ReVanced/revanced-patches/-/merge_requests/6956
- **Actual map rendering:** GmsCore PR → https://github.com/ReVanced/GmsCore/pull/370

**On the fix itself:**  
it's a pure bytecode neutralization, so it lives in the shared GmsCore-support patch, not the extension `GmsCoreSupport.java`.  
The existing `googlePlayUtilityMethod?.returnEarly(0)` only covers `GoogleApiAvailabilityLight`.  
Photos' bundled Maps SDK instead gates on `GooglePlayServicesUtilLight.isGooglePlayServicesAvailable(Context, int)` - a different class, matched by the same `"Google Play Services not available"` string you already use for `serviceCheckMethod`.  
I added a null-safe fingerprint for that `(Context,int)I` overload and return it early with `0`.

**On your question (does the map actually work, or does it just hide the error?)** — confirmed on-device, and it's two separate things:

- **The crash fix** opens MapExplore/Places and correctly shows geo-located photos (pins/thumbnails), so it's genuinely working, not just suppressing the error.
- **Whether the basemap tiles draw** depends on GmsCore's renderer. Stock GmsCore uses keyed Mapbox/Stadia vector styles, so the map is blank without a valid key. The GmsCore PR adds a key-less OpenStreetMap raster fallback (only when no Mapbox/Stadia key is configured) plus two small renderer fixes — loading the native lib from the correct package under renamed installs, and skipping the lite-mode snapshotter for raster styles. With both merged, Photos renders a real map with zero crashes.

**Comment by ak800i** [comment]
Let me summarize:
There are 3 separate improvements here:
1. Google Photos don't crash - owned by reavanced patches. Fixed.
2. Map is rendered - owned by GmsCore. No PR is currently active, and would be blocked behind the existing effort https://github.com/microg/GmsCore/pull/3401
3. Heatmap overlay is rendered - https://github.com/microg/GmsCore/pull/3481 or https://github.com/microg/GmsCore/pull/3595
